core/hybrisadaptor.cpp

```diff
@@ -391,23 +402,55 @@ void HybrisManager::initManager()
 
 HybrisManager::~HybrisManager()
 {
-    cleanup();
-
-#ifdef USE_BINDER
-    if (m_serviceManager) {
-        gbinder_servicemanager_unref(m_serviceManager);
-        m_serviceManager = NULL;
-    }
-#endif
+    /* This is exectuted after exiting main() function.
+     * No actions that need core application, binder ipc,
+     * android hal libraries, etc should be made.
+     */
 }
 
 void HybrisManager::cleanup()
 {
+    /* Stop any sensors that are active
+     */
+
     sensordLogD() << "stop all sensors";
     foreach (HybrisAdaptor *adaptor, m_registeredAdaptors.values()) {
         adaptor->stopSensor();
     }
 
+    /* Stop reacting to async events
+     */
+
+#ifdef USE_BINDER
+    gbinder_remote_object_remove_handler(m_remote, m_deathId);
+    m_deathId = 0;
+
+    if (m_pollTransactId) {
+        gbinder_client_cancel(m_client, m_pollTransactId);
+        m_pollTransactId = 0;
+
+        // The above code just marks down pending POLL transaction as
+        // to be cancelled later on when handler thread gets woken up.
+        //
+        // If we are exiting right after cleanup(), that is never going
+        // to happen and gbinder_ipc_exit() cleanup code blocks sensorfwd
+        // exit indefinitely.
+        //
+        // As a workaround, make a dummy synchronous POLL transaction to
+        // wake up the handler thread and get the cancellation processed.
```

Comment:
This doesn't sound right. If the asynchronous POLL transaction never gets completed and that causes `gbinder_ipc_exit()` to hang, why would a synchronous POLL transaction ever complete? From binder's point of view all transactions are synchronous. An asynchronous transaction in libgbinder is basically a (synchronous) transaction executed on a worker thread, whose results are getting asynchronously delivered back to the main thread.

Reply:
My understanding is that POLL behavior at sensor service side varies / args - it can return something immediately, or only after there is some changes to report. Similar dummy poll is already used for sanity checking during initialization. And having such thing done after cancellation does change behavior from "always hangs" to "have not seen it hang."

Reply:
Ah, that makes sense!